src/action.py

```diff
@@ -0,0 +1,366 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for taking the required actions to match the server state with the local state."""
+
+import logging
+import typing
+
+from . import exceptions, reconcile, types_
+from .discourse import Discourse
+
+DRAFT_NAVLINK_LINK = "<not created due to draft mode>"
+FAIL_NAVLINK_LINK = "<not created due to error>"
+DRAFT_MODE_REASON = "draft mode"
+NOT_DELETE_REASON = "delete_topics is false"
+
+
+def _absolute_url(url: types_.Url | None, discourse: Discourse) -> types_.Url | None:
+    """Get the absolute URL.
+
+    Args:
+        url The url to convert.
+        discourse: A client to the documentation server.
+
+    Returns:
+        The converted url or None if the url is None.
+    """
+    return discourse.absolute_url(url=url) if url is not None else None
+
+
+def _create(
+    action: types_.CreateAction, discourse: Discourse, draft_mode: bool, name: str
+) -> types_.ActionReport:
+    """Execute a create action.
+
+    Args:
+        action: The create action details.
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+        name: The charm name to prefix to the created pages title.
+
+    Returns:
+        A report on the outcome of executing the action.
+    """
+    logging.info("draft mode: %s, action: %s", draft_mode, action)
+
+    if action.content is None:
```

Comment:
This is a little difficult to follow

Reply:
I'll add a few comments, please let me know if that helps. Do note that some of these changes are part of the other PR, this PR will get much shorter once that one is merged